### DISCUSSION TITLE
Binding to Dictionary value using a string as key

### DISCUSSION THREAD

**Discussion by jasin** [AvaloniaUI/Avalonia#11871] [Q&A]
Binding to Dictionary value using a string as key? Or even better would be casting a enum as an int to pass as the key, but neither seem to be very easy to do. What's the better approach here?

This passes the syntax checker but causes an error in the initializeComponent
`Text="{CompiledBinding Accounts['somestring'], FallbackValue=0, StringFormat={}{0:C2}}" />`

XamlX.XamlParseException: Quote characters out of place Line 39, position 15.

I could use this, but I want the code to be expressive
`Text="{CompiledBinding Accounts[1000], FallbackValue=0, StringFormat={}{0:C2}}" />`

**Comment by timunie** [comment]
I guess a converter can help here.

**Reply by jasin** [reply]
I'm not sure I understand, doesn't the converter need to work on the bound data? I don't have bound data atm, I can't bind to a dictionary key that is a string. 

Are you saying I can do something like this
```
<TextBlock x:Name="tbChecking" Grid.Column="0" Text="General Checking" />
<TextBlock
        Grid.Column="1"
        Text="{CompiledBinding tbChecking, Converter={StaticResource NameOfConverter}, FallbackValue=0, StringFormat={}{0:C2}}" />
```

and use the string in the associated textblock to send as the value to the converter and then return the value of the Dictionary associated Key? Is there an example showing usage of an enum?

This is unventured territory for myself. While the above approach seems theoretical possible, it also seems fragile

**Reply by timunie** [reply]
okay I found this works for an enum-key as well as for a string: 

```xml
<TextBlock Text="{Binding Accounts[Key1]}" />
```

if you need more control, use a converter: 
```xml
<TextBlock Text="{Binding Accounts, Converter={StaticResource MyConverter}, ConverterParameter='Key1'}" />

<!-- with --> 

<Application.Resources>
    <conv:DictionaryEntryConverter x:Key="MyConverter" />
</Application.Resources>
```

```cs
public class DictionaryEntryConverter : IValueConverter
{
    public object? Convert(object? value, Type targetType, object? parameter, CultureInfo culture)
    {
        if (value is IDictionary dictionary)
        {
            return dictionary[parameter!];
        }

        return "value was null";
    }

    public object? ConvertBack(object? value, Type targetType, object? parameter, CultureInfo culture)
    {
        throw new NotSupportedException();
    }
}
```

**Comment by jasin** [comment]
> okay I found this works for an enum-key as well as for a string:

as long as the string doesn't contain spaces or other special characters. I was unable to find any sort of escaping etc. that did work for those types of strings.

**Reply by timunie** [reply]
converter should work with whitespaces as well. It can be used anywhere.
